### PR TITLE
Feature/messages

### PR DESCRIPTION
**Pull request by Sekir0**
# Description

Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change.

- Implementing messages functionality
- Cover all by API tests

## Type of change

Select single relevant option.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# Checklist:

- [x] I have performed a self-review of my code
- [x] My code follows the style guidelines and my changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: GhostNetwork.Gateway.Api/Messages/MessagesController.cs [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    [HttpPut("{messageId}")]
+    [ProducesResponseType(StatusCodes.Status204NoContent)]
+    [ProducesResponseType(StatusCodes.Status403Forbidden)]
+    [ProducesResponseType(StatusCodes.Status404NotFound)]
+    public async Task<ActionResult<Message>> UpdateAsync([FromRoute] string chatId, [FromRoute] string messageId, [FromBody] UpdateMessageModel model)
+    {
+        var chat = await chatStorage.GetByIdAsync(chatId);
+        var message = await messageStorage.GetByIdAsync(chatId, messageId);
+
+        if (chat is null || message is null)
```
[the remark this answers is not in the document]
Comment (volodymyr-borodin) [reply GhostNetwork.Gateway.Api/Messages/MessagesController.cs:122]: same for delete

File: GhostNetwork.Gateway.Infrastructure/ChatStorage.cs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    private static string GetCursorHeader(IApiResponse response)
+    {
+        return !response.Headers.TryGetValue("X-Cursor", out var headers)
+            ? default
+            : headers.FirstOrDefault();
+    }
+
+    private static Chat ToGatewayChat(GhostNetwork.Messages.Model.Chat chat)
```
Comment (volodymyr-borodin) [inline GhostNetwork.Gateway.Infrastructure/ChatStorage.cs:78]: use explicit operator for simple convertions

File: GhostNetwork.Gateway.Api/Messages/MessagesController.cs [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    /// <summary>
+    /// Get message by id.
+    /// </summary>
+    /// <param name="chatId">Chat id.</param>
+    /// <param name="messageId">Message id.</param>
+    /// <returns>Message.</returns>
+    [HttpGet("{messageId}")]
+    [ProducesResponseType(StatusCodes.Status200OK)]
+    [ProducesResponseType(StatusCodes.Status404NotFound)]
+    public async Task<ActionResult<Message>> GetByIdAsync([FromRoute] string chatId, [FromRoute] string messageId)
```
Comment (volodymyr-borodin) [inline GhostNetwork.Gateway.Api/Messages/MessagesController.cs:71]: What's the use case for this endpoint?

File: GhostNetwork.Gateway.Api/Messages/MessagesController.cs [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+using System.Threading.Tasks;
+using GhostNetwork.Gateway.Messages;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using Swashbuckle.AspNetCore.Filters;
+
+namespace GhostNetwork.Gateway.Api.Messages;
+
+[Route("{chatId}/messages")]
```
Comment (volodymyr-borodin) [inline GhostNetwork.Gateway.Api/Messages/MessagesController.cs:13]: don't you forget some part of the route?

File: GhostNetwork.Gateway.Api/Messages/MessagesController.cs [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+    /// <summary>
+    /// Create message.
+    /// </summary>
+    /// <param name="chatId">Chat id.</param>
+    /// <param name="model">Message model.</param>
+    /// <returns>Message.</returns>
+    [HttpPost]
+    [ProducesResponseType(StatusCodes.Status201Created)]
+    [ProducesResponseType(StatusCodes.Status404NotFound)]
+    public async Task<ActionResult<Message>> CreateAsync([FromRoute] string chatId, [FromBody] CreateMessage model)
```
Comment (volodymyr-borodin) [inline GhostNetwork.Gateway.Api/Messages/MessagesController.cs:93]: Please rename to `SendAsync`

File: GhostNetwork.Gateway.Infrastructure/ChatStorage.cs [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+    public async Task<DomainResult> UpdateAsync(string id, string name, IEnumerable<Guid> participants)
+    {
+        try
+        {
+            await chatsApi.UpdateAsync(id, new UpdateChatModel(name, participants.ToList()));
+        }
+        catch (ApiException)
+        {
+            return DomainResult.Error("API error!");
```
Comment (volodymyr-borodin) [inline GhostNetwork.Gateway.Infrastructure/ChatStorage.cs:60]: Expose error details

File: GhostNetwork.Gateway.Api/Chats/ChatsController.cs [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    /// Update chat.
+    /// </summary>
+    /// <param name="id">Chat id.</param>
+    /// <param name="model">Update chat model.</param>
+    [HttpPut("{id}")]
+    [ProducesResponseType(StatusCodes.Status404NotFound)]
+    [ProducesResponseType(StatusCodes.Status204NoContent)]
+    [ProducesResponseType(StatusCodes.Status403Forbidden)]
+    [ProducesResponseType(StatusCodes.Status400BadRequest)]
+    public async Task<ActionResult> UpdateAsync([FromRoute] string id, [FromBody] UpdateChat model)
```
Comment (volodymyr-borodin) [inline GhostNetwork.Gateway.Api/Chats/ChatsController.cs:100]: Add validation (not more than 10 participants) and Chat name length from 5 to 256 chars

**Review by volodymyr-borodin — CHANGES_REQUESTED** [review]

**Review by volodymyr-borodin — APPROVED** [review]

**Comment by volodymyr-borodin** [comment]
There are 2 comments left. Others marked as resolved

**Comment by Sekir0** [comment]
> There are 2 comments left. Others marked as resolved

But GetById endpoint is already removed from controller
And participants validation is in MessageValidator.cs

File: GhostNetwork.Gateway/Messages/MessageValidator.cs [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+        var chat = await chatStorage.GetByIdAsync(context.ChatId);
+
+        if (chat is null)
+        {
+            errors.Add(new DomainError("Chat not found"));
+        }
+        else if (chat.Participants.All(x => x.Id != context.AuthorId))
+        {
+            errors.Add(new DomainError("You is not a participant of the chat!"));
```
Comment (volodymyr-borodin) [inline GhostNetwork.Gateway/Messages/MessageValidator.cs:36]: You is

**Review by volodymyr-borodin — APPROVED** [review]

**Comment by volodymyr-borodin** [comment]
You did a great job

**Pull request merged**